### ISSUE TITLE
Cmake-js support

### ISSUE THREAD

**Issue by aminya** [prebuild/prebuildify#49]
prebuild supports cmake-js, but prebuildify does not. It would be good if they both are able to work with cmake-js.

**Comment by shahneil** [comment]
Any updates?

**Comment by aminya** [comment]
@shahneil yes! Please use cmake-ts
https://www.npmjs.com/package/cmake-ts

**Comment by Julusian** [comment]
@thegecko I would suggest doing something like what I have in a cmake-js based project. https://github.com/Julusian/node-jpeg-turbo/blob/main/.github/workflows/node.yaml
It will require some adjusting as it is written for https://github.com/julusian/pkg-prebuilds instead of https://github.com/prebuild/node-gyp-build, but I expect it should translate across without too much issue.

There is a 'mistake' in that workflow of mine currently, with it not using docker for the normal linux prebuild, but that is easily fixable to ensure it ends up using a sensible glibc.

**Comment by Freed-Wu** [comment]
https://github.com/julusian/pkg-prebuilds does not only support cmake, but also xmake [example](https://github.com/tonyfettes/coc-rime)

**Comment by segevfiner** [comment]
With prebuild not being deprecated, and this not supporting cmake-js, I'm stuck between a rock and a hard place...

**Comment by vweevers** [comment]
Couple of options:

- You can take this old PR - https://github.com/prebuild/prebuildify/pull/7. Rebase it on top of latest `prebuildify` and publish a forked package. It's unlikely to land here (no interest from core contributors I think) but the approach of that PR is clean and simple.
- Alternatively use @Julusian's `pkg-prebuilds`. I'm not familiar but comments above say it works for cmake-js.

**Comment by segevfiner** [comment]
Gotta love how bad the maintenance status of packages is in the npm ecosystem... So many poorly to unmaintained packages with no clean replacement... Constant breaking changes, projects replacing themselves and so on... Neverending tech debt on top never ending tech debt...

It's all open source, so no one really to complain to about unmaintained stuff, its just frustrating that it happens to stuff people use instead of trying to handover the project... And just choosing to leave it unmaintained or outright archive it, which is disruptive and really annoying...

**Comment by vweevers** [comment]
It can be frustrating, I hear you. It did not come out of the blue though. The earliest signal was in 2021 (https://github.com/prebuild/prebuild-install/commit/5ee1a2fdb974118a372c8e7b9508d89e89fbdf34) and the deprecation started 2 months ago. And ultimately, the decision to deprecate (even if abrupt) is better for the ecosystem than keeping it unmaintained. The packages can still be used btw, and as mentioned, alternatives are available (that IMO having been pushing the ecosystem in a good direction). Using open source software means subscribing to its risks.

All in all, sounds pretty healthy to me.

**Comment by thegecko** [comment]
> Gotta love how bad the maintenance status of packages is in the npm ecosystem... 

At least its not python packages! (ducks)

> Alternatively use @Julusian's pkg-prebuilds. I'm not familiar but comments above say it works for cmake-js.

Used by `webbluetooth` very successfully:

https://github.com/thegecko/webbluetooth/blob/main/.github/workflows/prebuild.yml
https://github.com/thegecko/webbluetooth/blob/main/package.json#L35

**Comment by segevfiner** [comment]
> > Gotta love how bad the maintenance status of packages is in the npm ecosystem...
> 
> At least its not python packages! (ducks)
> 

Off-topic: Try Golang packages, even more fun!

**Comment by aminya** [comment]
CMake-ts is an all-in-one tool that covers the functionalities provided by prebuildify, cmake-js, etc. Give it a try, it's been used by zeromq.js in production.

https://github.com/EmbeddedEnterprises/cmake-ts

**Comment by segevfiner** [comment]
> CMake-ts is an all-in-one tool that covers the functionalities provided by prebuildify, cmake-js, etc. Give it a try, it's been used by zeromq.js in production.
> 
> https://github.com/EmbeddedEnterprises/cmake-ts

Interesting, the example CMakeLists.txt looks very opinionated though... Not sure what parts of it are required vs what parts are example optional customizations.

**Comment by Julusian** [comment]
cmake-js and pkg-prebuilds both aren't going anywhere. 

pkg-prebuilds does work just as well with node-gyp too, I have a mix of libraries using it.
cmake-js is hanging in there, doesnt need all that much attention, but gets some when needed. I have a large redesign in the works, but need to figure out how to do a large breaking change in the least annoying way.

For a quick bit of history, I created pkg-prebuilds a few years ago as an alternative to node-gyp-build as I didn't like some of the decisions taken in node-gyp-build (largely 'poor' support for webpacked/esbuild bundled projects). So I went my own way with a small and opinionated tool

I keep forking and maintaining abandoned native modules (a couple I was added to instead), so my need isnt going away anytime soon

**Comment by aminya** [comment]
> > CMake-ts is an all-in-one tool that covers the functionalities provided by prebuildify, cmake-js, etc. Give it a try, it's been used by zeromq.js in production.
> > 
> > https://github.com/EmbeddedEnterprises/cmake-ts
> 
> Interesting, the example CMakeLists.txt looks very opinionated though... Not sure what parts of it are required vs what parts are example optional customizations.

That's an example of a production CMake that uses vcpkg (for cross platform dependencies). If you don't need that, you just need the last few lines. 
Feel free to contribute a minimal example.

**Comment by aminya** [comment]
Closing this as CMake-ts covers this feature and more.

https://github.com/EmbeddedEnterprises/cmake-ts

**Comment by segevfiner** [comment]
It doesn't look like it comes with full prebuilds infrastructure, parts of it seem like building and merging all the prebuilds into one artifact are hard coded in GitHub actions in the main project that uses this, AKA zeromq.js

Which is quite evident in the cmake-ts code which just literally says zeromq.js in places:
https://github.com/EmbeddedEnterprises/cmake-ts/blob/f116799b1e21cdcfafaefcc12785c784a6987f3c/src/loader.ts#L36-L42

**Comment by aminya** [comment]
Yes, the artifacts are included in the npm package itself so it is self contained similar to prebuldify for efficiency and reliability. 

Check out the readme of prebuldify:
> The irony of this approach is that it is faster to download all prebuilt binaries for every platform when they are bundled than it is to download a single prebuilt binary as an install script

**Comment by segevfiner** [comment]
Yeah, but its just not clear how to produce such a package using cmake-ts, versus how easy it is using prebuildify...

**Comment by aminya** [comment]
`npx cmake-ts build` on all platforms and merge the folders. Feel free to open an issue there and contribute more docs to the repo if it's not clear yet. The discussion is out of the scope of this issue.
